feat: generate dynamic docs header links

ShubhamOulkar opened the pull request:
#### New design
|light|dark|
|-----|----|
|<img width="284" height="31" alt="Screenshot 2025-09-14 084145" src="https://github.com/user-attachments/assets/115d9253-551f-4d85-b6ec-4875dc5247de" />|<img width="285" height="33" alt="Screenshot 2025-09-14 084157" src="https://github.com/user-attachments/assets/1a5cc633-9714-4be9-8ea6-1f852bbb5b3b" />|


<!--
Thank you for your pull request. Please provide a description and 
note the Certificate of Origin below. 

-->

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

ShubhamOulkar committed:
add header links

ShubhamOulkar committed:
fix redeclaration

ShubhamOulkar committed:
fix design

ShubhamOulkar committed:
add margin and font size

ShubhamOulkar committed:
change link content to hash

Signed-off-by: Shubham Oulkar <oulkarshubhu@gmail.com>

ShubhamOulkar committed:
remove outline and add color to hash

Signed-off-by: Shubham Oulkar <oulkarshubhu@gmail.com>

ShubhamOulkar committed:
Merge branch 'gh-pages' into generate-header-links

ShubhamOulkar committed:
correct aria-label value

jonchurch commented on css/style.css:
```diff
@@ -1116,6 +1116,26 @@ h2 a {
+}
+
+.heading-link {
+  margin-left: 0.3em;
```
```suggestion
  padding-inline: 0.3em;
```

I prefer padding for this, as it expands the clickable area to include the spacing, vs margin which adds spacing that is not clickable

ShubhamOulkar replied:
I think if we do font size inherit, padding will grow even bigger links especially on h2 headers.

ShubhamOulkar replied:
This looks good 👍
<img width="235" height="56" alt="image" src="https://github.com/user-attachments/assets/c387f19b-1461-4f24-ad1c-ffd4f324831d" />

jonchurch commented on css/style.css:
```diff
@@ -1116,6 +1116,26 @@ h2 a {
+
+.heading-link {
+  margin-left: 0.3em;
+  font-size: 1rem;
```
```suggestion
```

Leaving the font-size to inherit from the contained header will let the anchor match the heading size, otherwise you get mismatches like this:

<img width="259" height="169" alt="Image" src="https://github.com/user-attachments/assets/323882db-090f-43ae-8d65-88b1b7a59299" />

ShubhamOulkar replied:
yes I will do that

jonchurch commented on css/style.css:
```diff
@@ -1116,6 +1116,26 @@ h2 a {
+/* h2 -> h3 header links */
+.heading-link::after {
+  content: '#';
+  color: var(--card-fg);
```
```suggestion
```

If taking the nit https://github.com/expressjs/expressjs.com/pull/2055/files#r2370267450 we can drop color from here, or set color to inherit here. Both work, but styling the link is slightly more centralized.

jonchurch commented on css/style.css:
```diff
@@ -1116,6 +1116,26 @@ h2 a {
+
+h2:hover .heading-link,
+h3:hover .heading-link,
+.heading-link:is(:hover, :focus, :focus-visible) {
```
```suggestion
:target .heading-link,
.heading-link:is(:hover, :focus, :focus-visible) {
```

We could also show the anchor persistently when a user arrives via deep link into the content, or has clicked on a heading. Can sometimes help you orient after landing on a page.

This is purely stylistic feedback

ShubhamOulkar replied:
This is great, I have completely forgotten this. Thank you 💖

jonchurch commented on css/style.css:
```diff
@@ -1116,6 +1116,26 @@ h2 a {
+.heading-link {
+  margin-left: 0.3em;
+  font-size: 1rem;
+  text-decoration: none;
```
```suggestion
  text-decoration: none;
  color: inherit;
```

Nit: We can move styling of the pseudo element (the `::after` with `#`) color up a level. so we style the `a` itself to inherit from the header, and the pseudo element will inherit from there.

(will bork anything where the header is not the one providing a color, but that's useful in its own way)

ShubhamOulkar replied:
yes

jonchurch commented on js/copycode.js:
```diff
@@ -1,4 +1,5 @@
 const codeBlocks = document.querySelectorAll("pre:has(code)");
+const headers = document.querySelectorAll('h2[id],h3[id]');
```
This is broad enough that it is also selecting our sidebar which has an h3:

<img width="259" height="169" alt="Image" src="https://github.com/user-attachments/assets/0e8d9848-7381-4de0-8ea8-a7e3ae4094ae" />


We can scope it down to only select headers which appear inside the `<main>` tag
```suggestion
const headers = document.querySelectorAll('main :is(h2,h3)[id]');
```

jonchurch commented on js/copycode.js:
```diff
@@ -13,6 +14,15 @@ codeBlocks.forEach((block) => {
 });
 
+// add heading links
+for (const heading of headers) {
```
```suggestion
for (const heading of headers) {
    if (heading.querySelector('.heading-link')) continue;
```
Nit: A little paranoid but this will guard against the script running twice somehow by accident or hot reloading or something, preventing us from adding the elements twice.

ShubhamOulkar replied:
I want thinking of implementing lazy attach links using  `IntersectionObserver` or `MutationObserver`. WDYT?

> Is there a reason why this is living inside the copycode.js script file?

Copy code and header link share same logic and needs asynchronous implementation.

ShubhamOulkar replied:
The DOM is reloaded completely on each hot reload, so duplicates won’t occur. Still adding this check as a defensive measure in case the script ever re-runs

jonchurch reviewed (changes requested):
For some reason I thought we already had these, or had them in a PR, but I can't find any evidence for that.

Great idea. 

Do note though, this will add these anchor links to `h2, h3` on **all pages** of the site!

I didn't check all pages to see if it is messed up anywhere, but I caught some bugs on the docs page at least.

Is there a reason why this is living inside the `copycode.js` script file? it could move to the `app.js` file inside the jquery DOMContentLoaded callback at the top of the file

ShubhamOulkar committed:
Scope header selection to main section

Co-authored-by: Jon Church <me@jonchurch.com>

ShubhamOulkar committed:
add target highliting to header links

Co-authored-by: Jon Church <me@jonchurch.com>

ShubhamOulkar committed:
inherit color and font size from parent headers

ShubhamOulkar committed:
expand link clickable area

Co-authored-by: Jon Church <me@jonchurch.com>

ShubhamOulkar commented:
I have addressed all reviews. Please take a look.

> it could move to the `app.js` file inside the jquery

Yes, I will move this code to app.js. Just a reminder, jQuery is migrated to web APIs. So no more jquery.

> Do note though, this will add these anchor links to `h2, h3` on **all pages** of the site!

Yes, this will add anchor on all pages. Here are some issues that are affecting this PR https://github.com/expressjs/expressjs.com/pull/2055#discussion_r2370334622 and #2056. First is fixed here itself, #2056 needs to be fixed.

ShubhamOulkar commented:
bump @jonchurch

ShubhamOulkar committed:
Merge branch 'gh-pages' into generate-header-links

ShubhamOulkar committed:
Merge branch 'gh-pages' into generate-header-links

ShubhamOulkar committed:
add defensive check on heading

Co-authored-by: Jon Church <me@jonchurch.com>

bjohansebas commented on js/copycode.js:
Why is this new logic here instead of in app.js?

ShubhamOulkar commented:
@bjohansebas, I think astro can do it in better way. You can close this PR if not necessary.

ShubhamOulkar commented:
astro can do it in better way.
